File: v2/golang/client/upload_types.go
package client

// upload_types.go — one domain: the two response shapes of the file-upload
// route (POST /api/v1/upload and GET /api/v1/upload/{id}/status).
//
// Split out of types.go on 2026-09-14 (types.go was past the ~300-line house
// cut) while both structs were being reconciled with the handler.

// UploadResult is the HTTP 202 acceptance envelope of POST /api/v1/upload.
//
// Field set is the handler's fixed map, server/handler/upload.go:109-119. There
// is NO `id` key on this route and there never was: the struct declared one
// until 2026-09-14 and it decoded to zero on every single upload, which is why
// UploadID() needed a fallback that could never fire.
//
// Junior Tip [202 means accepted, not ingested]: the record exists but its
// content is still being extracted. Status here is the status AT ACCEPTANCE.
// Poll UploadStatus (or WaitForUpload) for the terminal one.
type UploadResult struct {
	Message      string `json:"message"`
	RecordID     int64  `json:"record_id"`
	UUID         string `json:"uuid,omitempty"`
	Filename     string `json:"filename,omitempty"`
	MIME         string `json:"mime,omitempty"`
	MIMEDetected string `json:"mime_detected,omitempty"`
	Extension    string `json:"extension,omitempty"`
	SizeBytes    int64  `json:"size_bytes,omitempty"`
	Status       string `json:"status,omitempty"`
}

// UploadID returns the server record id used for UploadStatus polling.
//
// Junior Tip [why this is not just a field read]: it used to fall back to a
// phantom `ID` field when RecordID was zero. That fallback made a zero
// RecordID — which only happens when the response failed to decode — look like
// a legitimate id of 0, and the caller then polled /upload/0/status forever.
// One source, no fallback: a zero return now means "the accept response did not
// carry a record_id", which is a real failure the caller must see.
func (uploadResult UploadResult) UploadID() int64 {
	return uploadResult.RecordID
}

// UploadStatusResult describes the processing status of a file upload.
//
// Field set is the handler's fixed 7-key map, server/handler/upload.go:220-236:
// record_id, uuid, status, type, summary, metadata, completed. Nothing else,
// ever.
//
// Junior Tip [failure arrives as status, never as an error field, 2026-09-14]:
// this struct used to declare ID, Filename and Error. The server sends none of
// the three. Error in particular was load-bearing in the wrong direction — the
// wait loop treated `Error != ""` as a terminal condition, a branch that could
// not fire, so a failed ingest was only ever caught by status == "failed"
// sitting next to it. A failed ingest is reported through `status`, and only
// through `status`.
type UploadStatusResult struct {
	RecordID  int64  `json:"record_id"`
	UUID      string `json:"uuid,omitempty"`
	Status    string `json:"status"` // "processing", "completed", "failed", "saved"
	Type      string `json:"type,omitempty"`
	Completed bool   `json:"completed"`
	Summary   string `json:"summary,omitempty"`
	Metadata  string `json:"metadata,omitempty"`
}

File: v2/typescript/src/uploadTypes.ts
/**
 * AnhurDB TypeScript SDK — what the two UPLOAD endpoints ANSWER.
 *
 * `POST /api/v1/upload` (202 Accepted) and `GET /api/v1/upload/{id}/status`.
 * Its own file because `types.ts` is long past the ~300-line house cut and
 * house law forbids growing a file already over it; upload is a real domain
 * with one handler (`AnhurDB/server/handler/upload.go`) owning both shapes.
 *
 * @module
 */

/**
 * Result of `POST /api/v1/upload` — the file was ACCEPTED, not processed.
 *
 * Mirrors the fixed nine-key map at `handler/upload.go:109-119` exactly. The
 * status code is 202: ingestion is asynchronous, so poll `uploadStatus()` /
 * `waitForUpload()` with `record_id` to learn what became of the file.
 *
 * Junior Tip [there is no `id`, and there never was — 2026-09-14]: this
 * interface used to declare `id?: number`, and the handler has never emitted
 * it. The key for polling is `record_id`. A phantom optional is the worst
 * kind of lie a type can tell: `upload.id ?? upload.record_id` compiles, reads
 * like a careful fallback, and the first branch is dead code that no test can
 * ever reach. Every field below is a key in that map; nothing in that map is
 * missing here. When you add one, open the handler first.
 *
 * Every field is optional because an OSS/older deployment may answer with a
 * subset — but a field absent from the handler does not belong here at all.
 */
export interface UploadResult {
  /** Human-readable ack, currently `"File accepted for processing"`. */
  message?: string;
  /** The file record's id — the value `uploadStatus()` takes. */
  record_id?: number;
  /** Session UUID the file was filed under (`result.SessionUUID`). */
  uuid?: string;
  /** Original filename as the server stored it. */
  filename?: string;
  /** MIME type the server settled on. */
  mime?: string;
  /** MIME type SNIFFED from the bytes; today the handler sends the same value. */
  mime_detected?: string;
  /** Extension the server derived (e.g. `"pdf"`). */
  extension?: string;
  /** Size of the accepted payload in bytes. */
  size_bytes?: number;
  /** Initial ingest status, e.g. `"processing"`. */
  status?: string;
}

/**
 * Result from upload status polling — `GET /api/v1/upload/{id}/status`.
 *
 * Mirrors `UploadHandler.UploadStatus` in
 * `AnhurDB/server/handler/upload.go` EXACTLY: the handler builds a fixed map of
 * seven keys and never adds another. Every key below exists there; nothing that
 * exists there is missing here.
 *
 * Junior Tip [a declared field the server never sends is worse than a missing
 * one]: this interface used to declare `filename`, `error`, `record_ids` and a
 * bare `id`. None of them are ever emitted — the handler answers
 * `record_id/uuid/status/type/summary/metadata/completed`. The cost was not
 * cosmetic: `waitForUpload` used `Boolean(payload.error)` as one of its
 * TERMINAL conditions, so that branch was unreachable code that read as a
 * safety net. A failed ingest is reported through `status`, and only through
 * `status`. When you add a field here, open the handler first; the type is a
 * claim about the server, not a wish list.
 */
export interface UploadStatusResult {
  /** The file record's id (server key is `record_id`, never `id`). */
  record_id?: number;
  /** Stable record uuid. */
  uuid?: string;
  /** "processing", "completed", "saved", or "failed". */
  status: string;
  /** Always `"file"` — the handler rejects anything else with HTTP 400. */
  type?: string;
  /** Server-computed: true when `status` is `completed` or `saved`. */
  completed?: boolean;
  summary?: string;
  /** Raw metadata JSON string, exactly as stored. */
  metadata?: string;
}
